XLS-40: DID Specification (#136)
XRPLF/XRPL-Standards: XLS-0040d-decentralized-identity/README.md

- This proposal is NOT intended as a competing decentralized identity standard; it aims to leverage existing (and emerging) decentralized identity standard proposed by W3C.
  
- The proposal is motivated to enable **individual users** to create, and manage their decentralized identifiers while having complete control over the private keys and contents of the identity object.
 
# 3. New On-Ledger Objects and Transactions
We propose one new ledger object and two new transaction types. 
- **DID** is a new ledger object that is unique to an XRPL account and may contain a reference to, or the hash of, or the [W3C DID document](https://w3c-ccg.github.io/did-primer/#did-documents) itself associated with the corresponding DID.
- **DIDSet** is a new transaction type used to perform the following two operations:
  -  `Create` the new ledger object `DID` with `URI`, and/or `Data` fields, thus adding the required reserve towards the account that initiated the transaction.

>   -  `Update` the mutable `URI` and `Data` fields of the `DID` object.

Reviewer: This section needs to be updated to include `DIDDocument`

XRPL DID MUST be unique within the XRPL network. 

`xrpl-specific-idstring` is the `AccountID` or the hex of `master public key` of the `DID` object's account. See [this](https://xrpl.org/accounts.html#address-encoding) for more details on XRPL public keys.

Example:
A valid DID for an XRPL network may be
```
did:xrpl:1:rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh
```
or
```
did:xrpl:1:0330E7FC9D56BB25D6893BA3F317AE5BCF33B3291BD63DB32654A313222F7FD020
```
The above DIDs are case-insensitive and will resolve to the same `DID` object.

# 5. CRUD Operations
In this section we outline the following four CRUD operations for ``did:xrpl`` method. 
- `Create`
- `Read`
- `Update` 
- `Delete` 

## 5.1. DIDSet Transaction 
The proposal introduces a new transaction type: **`DIDSet`** that can be used to:
- `Create` a new `DID` object,
- `Update` the existing `DID` object.   

### 5.1.1 Example **`DIDSet`** JSON
```
{
  "TransactionType": "DIDSet",
  "Account": "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
  "Fee": "10",
  "Sequence": 391,
  "URI": "697066733A2F2F62616679626569676479727A74357366703775646D37687537367568377932366E6634646675796C71616266336F636C67747179353566627A6469",

>   "Data": ,

Reviewer: ```suggestion
  "Data": "",
```

 
### Example
Given the DID **did:xrpl:1:rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn**, do the following:
- Retrieve `rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn`
- Perform the `account_objects` method request to retrieve the contents of `DID` object:
```
{
    "method": "account_objects",
    "params": [
        {
            "account": "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
            "ledger_index": "validated",
            "type": "did"
        }
    ]
}
```

A sample response to the above query might look like this:

```
200 OK
{
    "result": {
        "account": "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
        "account_objects": [
            {
                "URI": "697066733A2F2F62616679626569676479727A74357366703775646D37687537367568377932366E6634646675796C71616266336F636C67747179353566627A6469"
            }
        ],
        "validated": true
    }

> }

Reviewer: Here's an actual example output, if you want it:
```
{
  account: 'rDZdorj46soYhJKa26jm2cpJkgM3AfGhvu',
  account_objects: [
    {
      Account: 'rDZdorj46soYhJKa26jm2cpJkgM3AfGhvu',
      DIDDocument: '646F63',
      Data: '617474657374',
      Flags: 0,
      LedgerEntryType: 'DID',
      OwnerNode: '0',
      PreviousTxnID: '2052324F7A8F2122B7A66D35E75CBABC25CA7FEAE2C1806EB9F537220A03336F',
      PreviousTxnLgrSeq: 4,
      URI: '6469645F6578616D706C65',
      index: 'A6F0C71B1E6EB186ED20906E655F69C693AE3C9D4D19D2110DCB85DBE1E87216'
    }
  ],
  ledger_hash: '95AD8063A729FDE24EFD4A0423A26FD3F154A6DE6ECCDE99434C195D9C5BAB36',
  ledger_index: 4,
  validated: true
}
```

Author: Updated!

```

> Upon receiving this response, an IPFS address (CAS) can be retrieved from the `URI` field. This address can then be resolved to retrieve the corresponding DID Document.

Reviewer: This should note that the `DIDDocument` field might also contain the data directly.

Author: Updated!

There are several security and privacy considerations that implementers would want to take into consideration when implementing this specification.

## 6.1. Key Management

The entity which controls the private key associated with the `DID` object, i.e. the XRPL `Account` also effectively controls the reference to the DID Document which the DID resolves to. Thus great care should be taken to ensure that the private key is kept private. Methods for ensuring key privacy are outside the scope of this document.

## 6.2. DID Document Public Profile
The DID document or other data associated with the XRPL DID's `URI` and/or `Data`, and/or `DIDDocument` fields can contain any content, though it is recommended that it conforms to the W3C DID Document and Verifiable Credentials (VC) specification. As anchored DIDs on XRPL can be resolved by anyone, care should be taken to only update to resolve DID Documents and data which DO NOT expose any sensitive personal information, or information which one may not wish to be public. DID documents should be limited to verification methods and service endpoints, and SHOULD not store any personal information.

## 6.3. IPFS and Canonicity

> IPFS allows anyone to store content publicly on the nodes in a distributed network. A common misconception is that anyone can edit content, however the content-addressability of IPFS means that this new edited content will have a different address to the original. While any entity can copy a DID Document anchored with an XRPL account's `URI`, they cannot change the document that a DID resolves to via the XRPL `account_objects` resolution unless they control the private key which created the corresponding `DID` object.

Reviewer: ```suggestion
IPFS allows anyone to store content publicly on the nodes in a distributed network. A common misconception is that anyone can edit content, however the content-addressability of IPFS means that this new edited content will have a different address to the original. While any entity can copy a DID Document anchored with an XRPL account's `URI` or `DIDDocument`, they cannot change the document that a DID resolves to via the XRPL `account_objects` resolution unless they control the private key which created the corresponding `DID` object.
```

Author: Updated!

For more, see [§ 10. Privacy Considerations](https://www.w3.org/TR/did-core/#privacy-considerations) in did-core. 

# 7. Implementations
A proposed implementation is in progress here: https://github.com/XRPLF/rippled/pull/4636/

# 8. Appendices
## Appendix 1. DID Document

A DID is associated with a DID document. A DID document contains the necessary information to cryptographically verify the identity of the DID subject. W3C defines the [core properties](https://www.w3.org/TR/did-core/#core-properties) in a DID document, that describe relationships between the DID subject and the value of the property. For example, a DID document could contain cryptographic public keys such that the DID subject can use it to authenticate itself and proves its association with the DID. Usually, a DID document can be serialized to JSON representation or JSON-LD representation (see [DIDs v1.0]).

Applications may choose one of the following for the DID document associated with a DID:
- Store a reference on the ledger in the `URI` field of `DID` object to the DID document stored in one or more parts on other decentralized storage networks such as IPFS or STORJ.

> - Store a minimal DID document in the `Data` field of the `DID` object.

Reviewer: ```suggestion
- Store a minimal DID document in the `DIDDocument` field of the `DID` object.
```

`URI` specifies a Universal Resource Identifier, a.k.a. URI that SHOULD point to the corresponding DID document or to the data associated with the DID. This field could be an HTTP(S) URL or IPFS URI. The `URI` field is NOT checked for validity, but the field is limited to a maximum length of 256 bytes. 

---
| Field Name   |     Required?      | JSON Type | Internal Type 
| -------------| :----------------: | :-------: | :-----------: 
| `Data`       |                    | `string`  |   `BLOB`    |

`Data` field SHOULD contain the public attestations of identity credentials associated with the DID. The `Data` field is NOT checked for validity, but the field is limited to a maximum length of 256 bytes. 

---
| Field Name   |     Required?      | JSON Type | Internal Type 
| -------------| :----------------: | :-------: | :-----------: 
| `DIDDocument`       |                    | `string`  |   `BLOB`    |

`DIDDocument` field SHOULD contain the <a href="https://www.w3.org/TR/did-core/#did-documents"> DID document </a> per W3C standards associated with the DID. The field is NOT checked for validity, and is limited to a maximum length of 256 bytes.

Note:

> - Either `URI`, `Data` or `DIDDocument` field MUST be present to create and update the `DID` object.

Reviewer: I thought this was just `URI` or `DIDDocument` that needed to be present.

Author: It should be either `URI`, `Data` or `DIDDocument` field.

Reviewer 2: Per @mvadari - she doesn't have the power to resolve this one, but agrees with the comment above from @aanchal4 .

> - All `URI`, `Data` and `DIDDocument` fields MUST be absent to delete the `DID` object.

Reviewer 2: Remind me why this must be true?

Reviewer: This isn't true in the implementation - in fact, you can't have a DID object without any of those fields. This doesn't make sense to me either, but happy to change if there's a reason.

Author: This is for a transaction, not for an object. I can rephrase it if it's confusing.

Reviewer: Maybe add another note that says something like "A valid DID object must have either a `URI` or `DIDDocument` field"?

Example:
A valid DID for an XRPL network may be
```
did:xrpl:1:rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh
```
or
```
did:xrpl:1:0330E7FC9D56BB25D6893BA3F317AE5BCF33B3291BD63DB32654A313222F7FD020
```
The above DIDs are case-insensitive and will resolve to the same `DID` object.

# 5. CRUD Operations
In this section we outline the following four CRUD operations for ``did:xrpl`` method. 
- `Create`
- `Read`
- `Update` 
- `Delete` 

## 5.1. DIDSet Transaction 
The proposal introduces a new transaction type: **`DIDSet`** that can be used to:
- `Create` a new `DID` object,
- `Update` the existing `DID` object.   

### 5.1.1 Example **`DIDSet`** JSON
```
{
  "TransactionType": "DIDSet",
  "Account": "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
  "Fee": "10",
  "Sequence": 391,
  "URI": "ipfs://bafybeigdyrzt5sfp7udm7hu76uh7y26nf4dfuylqabf3oclgtqy55fbzdi",
  "Data": ,
  "SigningPubKey":"0330E7FC9D56BB25D6893BA3F317AE5BCF33B3291BD63DB32654A313222F7FD020",
  ...

  }
```
## Create
If the `DID` object associated with the `Account` does not exist, 

>   - If both `URI` and `Data` fields are not present in the transaction, then transaction fails.

Reviewer 2: Should we also add `DIDDocument` here to align with https://github.com/XRPLF/XRPL-Standards/pull/136/files#diff-b0de935ca4c9ab9568454d8b2c3bd0d2637314702ece03008631d5f4f96e0bd3R104

>   - If both `URI` and `Data` fields are not present in the transaction, then transaction fails.

Reviewer 2: I think this should align with [this line](https://github.com/XRPLF/XRPL-Standards/pull/136/files#diff-b0de935ca4c9ab9568454d8b2c3bd0d2637314702ece03008631d5f4f96e0bd3R104)?

```suggestion
    - If `URI`, `Data` and `DIDDocument` fields are not present, then transaction fails.
```

Reviewer 3: thoughts @mvadari ?

Author: It should be - If either `URI`,`Data` or `DIDDocument` fields are not present, then the transaction fails.

Reviewer: Do you need all 3 if `Data` is usually just an attestation? Seems like only `URI` or `DIDDocument` is needed if you don't need an attestation.

Author: "If either `URI`, `Data` or `DIDDocument` fields are not present, then the transaction fails" means that for a transaction to be considered valid or successful, it must contain at least one of the following fields: `URI`, `Data`, or `DIDDocument`.

Reviewer: Sorry, yes, I was confusing the object and the transaction. The transaction needs one of the three, the object only needs `URI` or `DIDDocument`.

Reviewer 2: > The transaction needs one of the three,

This section _is_ for a transaction. @aanchal4 do you agree with this change? If so, please commit it.

---

| Field Name        |     Required?      | JSON Type | Internal Type 
| ----------------- | :----------------: | :-------: | :-----------: 
| `URI`             |                    | `string`  |   `BLOB`    |

`URI` specifies a Universal Resource Identifier, a.k.a. URI that SHOULD point to the corresponding DID document or to the data associated with the DID. This field could be an HTTP(S) URL or IPFS URI. The `URI` field is NOT checked for validity, but the field is limited to a maximum length of 256 bytes. 

---
| Field Name   |     Required?      | JSON Type | Internal Type 
| -------------| :----------------: | :-------: | :-----------: 
| `Data`       |                    | `string`  |   `BLOB`    |

`Data` field SHOULD contain the public attestations of identity credentials associated with the DID. The `Data` field is NOT checked for validity, but the field is limited to a maximum length of 256 bytes. 

---
| Field Name   |     Required?      | JSON Type | Internal Type 
| -------------| :----------------: | :-------: | :-----------: 
| `DIDDocument`       |                    | `string`  |   `BLOB`    |

> `DIDDocument` field SHOULD contain the <a href="https://www.w3.org/TR/did-core/#did-documents"> DID document </a> per W3C standards associated with the DID. The field is NOT checked for validity, and is limited to a maximum length of 256 bytes.

Reviewer: ```suggestion
`DIDDocument` field SHOULD contain the <a href="https://www.w3.org/TR/did-core/#did-documents"> DID document </a> per W3C standards associated with the DID. The `DIDDocument` field is NOT checked for validity, and is limited to a maximum length of 256 bytes.
```

### 4.1.2. The `DID` Object ID Format
`DID` object may contain public data associated with the XRPL account's identity. This requires the `DID` to be:
- Unique in the XRPL namespace, and 
- Uniquely associated with the XRPL account. 
 

> We compute the `DID` object ID, a.k.a., `DIDID`, as the `SHA-512Half` of the following values, concatenated in order:

Reviewer: It might be less confusing to use `index` instead of `ID` here to refer to the "object ID/index"

Author: We've consistently used the `ID` in our specifications and documentation. To maintain consistency, I propose that we keep it as is.

In the case of XRPL DID, the `read` operation facilitates the resolution to the associated `DID` object. Upon successful resolution, applications can retrieve and utilize the relevant `URI` and/or `Data`, and/or `DIDDocument` fields contained within the `DID` object:
 
Given the input DID for an account, follow these steps:

- Retrieve the xrpl-specific-idstring of the DID.

- If the xrpl-specific-idstring represents a public key:

  - Compute the Account ID using the method described [here](https://xrpl.org/accounts.html#address-encoding).

- Retrieve the contents of the corresponding `DID` object in its raw format using the XRPL's [`account_objects`]( https://xrpl.org/account_objects.html#) command. Use the `account` field to specify the Account ID to retrieve all objects owned by that account, including `DID` object in the raw ledger format. Or optionally use the `type` field to filter the results by `ledger_entry` type, i.e. `DID` to retrieve the contents of just the `DID` object.
 
### Example
Given the DID **did:xrpl:1:rDZdorj46soYhJKa26jm2cpJkgM3AfGhvu**, do the following:
- Retrieve `rDZdorj46soYhJKa26jm2cpJkgM3AfGhvu`

> - Perform the `account_objects` method request to retrieve the contents of `DID` object:

Reviewer: Thinking about this again, it would be better to do with `ledger_entry` and the param of `"did": "rDZdorj46soYhJKa26jm2cpJkgM3AfGhvu"` - there's no worry of dealing with pagination then.

Author: Updated the recommended method to `ledger_entry` and the `account_objects` method as an alternative.
